crates/egui/src/atomics/atomic_layout.rs

```diff
@@ -0,0 +1,392 @@
+use crate::{
+    Atomic, AtomicKind, Atomics, FontSelection, Frame, Id, IntoAtomics, Response, Sense,
+    SizedAtomic, SizedAtomicKind, Ui, Widget,
+};
+use ahash::{HashMap, HashMapExt};
+use emath::{Align2, NumExt, Rect, Vec2};
+use epaint::text::TextWrapMode;
+use epaint::Color32;
+
+/// Intra-widget layout utility.
+///
+/// Used to lay out and paint [`Atomic`]s.
+/// This is used internally by widgets like [`crate::Button`] and [`crate::Checkbox`].
+/// You can use it to make your own widgets.
+///
+/// Painting the atomics can be split in two phases:
+/// - [`AtomicLayout::allocate`]
+///   - calculates sizes
+///   - converts texts to [`Galley`]s
+///   - allocates a [`Response`]
+///   - returns a [`AllocatedAtomicLayout`]
+/// - [`AllocatedAtomicLayout::paint`]
+///   - paints the [`Frame`]
+///   - calculates individual [`Atomic`] positions
+///   - paints each single atomic
+///
+/// You can use this to first allocate a response and then modify, e.g., the [`Frame`] on the
+/// [`AllocatedAtomicLayout`] for interaction styling.
+pub struct AtomicLayout<'a> {
```

Comment:
This is a pretty chonky 96 bytes big. I wonder what the performance implications of this PR will be, both on native and web 😬